frame/crowdloan-rewards/src/lib.rs

```diff
@@ -0,0 +1,325 @@
+
+	impl<T: Config> Pallet<T> {
+		fn do_claim(
+			remote_account: RemoteAccountOf<T>,
+			reward_account: &T::AccountId,
+		) -> Result<T::Balance, DispatchError> {
+			ensure!(VestingBlockStart::<T>::exists(), Error::<T>::NotInitialized);
+			Rewards::<T>::try_mutate(remote_account, |reward| {
+				reward
+					.as_mut()
+					.map(|reward| {
+						let upfront_payment = T::InitialPayment::get().mul_floor(reward.total);
+						let should_have_claimed = {
+							let current_block = frame_system::Pallet::<T>::block_number();
+							// Current point in time
+							let vesting_point =
+								current_block.saturating_sub(VestingBlockStart::<T>::get());
+							if vesting_point >= reward.vesting_period {
+								// If the user is claiming when the period is over, he should
+								// probably have already claimed everything.
+								reward.total
+							} else {
+								let vesting_period = VestingBlockStart::<T>::get()
+									.safe_add(&reward.vesting_period)?;
+								let partition = T::VestingPartition::get();
+								// Current window, rounded to previous window.
+								let vesting_window = vesting_point - (vesting_point % partition);
+								// The user should have claimed the upfront payment + the vested
+								// amount until this window point.
+								let vested_reward = reward.total - upfront_payment;
+								upfront_payment +
+									(vested_reward
+										.saturating_mul(T::Convert::convert(vesting_window)) /
+										T::Convert::convert(vesting_period))
+							}
+						};
+						let available_to_claim = should_have_claimed - reward.claimed;
+						ensure!(
+							available_to_claim > T::Balance::zero(),
+							Error::<T>::NothingToClaim
+						);
+						T::Currency::mint_into(reward_account, available_to_claim)?;
+						(*reward).claimed += available_to_claim;
+						ClaimedRewards::<T>::mutate(|x| *x += available_to_claim);
+						Ok(available_to_claim)
+					})
+					.unwrap_or_else(|| Err(Error::<T>::NotAContributor.into()))
+			})
+		}
+	}
+
+	pub fn verify_relay<AccountId: Encode, RelayChainAccountId: Into<AccountId32>>(
+		prefix: &[u8],
+		reward_account: AccountId,
+		relay_account: RelayChainAccountId,
+		proof: MultiSignature,
+	) -> bool {
+		let wrapped_prefix: &[u8] = b"<Bytes>";
+		let wrapped_postfix: &[u8] = b"</Bytes>";
+		let mut msg = wrapped_prefix.to_vec();
+		msg.append(&mut prefix.to_vec());
+		msg.append(&mut reward_account.using_encoded(to_ascii_hex));
+		msg.append(&mut wrapped_postfix.to_vec());
+		proof.verify(&msg[..], &relay_account.into())
+	}
+
+	pub fn to_ascii_hex(data: &[u8]) -> Vec<u8> {
+		let mut r = Vec::with_capacity(data.len() * 2);
+		let mut push_nibble = |n| r.push(if n < 10 { b'0' + n } else { b'a' - 10 + n });
+		for &b in data.iter() {
+			push_nibble(b / 16);
+			push_nibble(b % 16);
+		}
+		r
+	}
+
+	pub fn ethereum_signable_message(prefix: &[u8], msg: &[u8]) -> Vec<u8> {
+		let mut l = prefix.len() + msg.len();
+		let mut rev = Vec::new();
+		while l > 0 {
```

Comment:
not sure, it is some eth standard? what if prefix1 + msg1 = prefix2 + msg2 possibl. is it safe? if rev is revision which is version, may be store v1, v2, as enum explicitly?

Reply:
does not feels as being some standard

Reply:
It's not a standard,  it's just a specific format for claiming, prefix+account_id in this case, I reused the claiming mechanism of  [polkadot](https://github.com/paritytech/polkadot/blob/master/runtime/common/src/claims.rs#L530)

Reply:
I mean the message header is standard

Reply:
can we reference code of https://github.com/paritytech/polkadot/blob/12df8e0777c0fa2ea92a19ae6d4f1d5bc3917bca/runtime/common/src/claims.rs?_pjax=%23repo-content-pjax-container#L110 instead of claiming ? I reference this for xcmp setup seems anyway.

Reply:
Yes I will reference the original code by documenting everything